### PR TITLE
Manage resources, constraint, and properties

### PR DESCRIPTION
**Pull request by OliPou**
This modification adds loops to configure cluster resources, constraint and properties through the ansible role

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -272,6 +347,31 @@ For cluster to get properly authorized it is expected that firewall is already c
...
 
+**Example playbook Resources configuration** .
+
+    - hosts: cluster
+      vars:
+        cluster_property:
+          - name: 'stonith-enabled'
+            value: 'false'
+          - name: 'no-quorum-policy'
+            value: 'ignore'
```
Comment (OndrejHome) [inline README.md]: Would it be possible to provide less controversial cluster properties here? 

The `stonith-enabled=false` and `no-quorum-policy=ignore` would render this cluster unsupportable on RHEL.
As a better example I would suggest `maintenance-mode=true` from the `pcs_property` examples doc that will create cluster in maintanence mode.

File: README.md [hunk 2]
```diff
@@ -272,6 +347,31 @@ For cluster to get properly authorized it is expected that firewall is already c
...
+            options: 'ip=192.168.1.150 cidr_netmask=24 meta migration-threshold=2 op monitor interval=20'
+        cluster_constraint_colocation:
+          - resource1: 'cluster_vip'
+            resource2: 'apache2'
+            score: 'INFINITY'
+        cluster_resource_defaults:
+          - name: 'failure-timeout'
+            value: '30'
+      roles:
+         - { role: 'ondrejhome.ha-cluster-pacemaker', cluster_name: 'apache-cluster', cluster_configure_fence_xvm: false }
```
Comment (OndrejHome) [inline README.md]: Would it be possible to remove `cluster_configure_fence_xvm: false` from here as well?

Reason is that it is not necessary for allowing the cluster resource configuration nor cluster property changes. My preference is on having rather shorter examples with only needed options to describe the new situations.

**Review by OndrejHome — CHANGES_REQUESTED** [review]
Hello @OliPou,

Thank you for using the ha-cluster-pacemaker role and for opening the PR to include support for configuring additional aspects of cluster such as resources and properties.

In additional to few comments/requests in code changes I have the following questions to this PR:
- On which system have you tested the changes? (Debian 10 only or also some other systems)
- Which Ansible version you have used to test changes here? I typically run tests on versions 2.8, 2.9 and 2.10 these days)
- What is the reason for changes in `README.md` on lines 75, 81, 89 and 116? It looks to me that it was most probably caused by editor trying to automatically "fix something".

Ondrej (온드레이)

File: README.md (hunk 1 above)
Comment (OliPou) [reply README.md]: Yes sure! I juste pickup one of my test. It should changed now

File: README.md (hunk 2 above)
Comment (OliPou) [reply README.md]: Yes sure! it should changed now

File: README.md (hunk 1 above)
Comment (OliPou) [reply README.md]: > Hello @OliPou,
> 
> Thank you for using the ha-cluster-pacemaker role and for opening the PR to include support for configuring additional aspects of cluster such as resources and properties.
> 
> In additional to few comments/requests in code changes I have the following questions to this PR:
> 
>     * On which system have you tested the changes? (Debian 10 only or also some other systems)
> 
>     * Which Ansible version you have used to test changes here? I typically run tests on versions 2.8, 2.9 and 2.10 these days)
> 
>     * What is the reason for changes in `README.md` on lines 75, 81, 89 and 116? It looks to me that it was most probably caused by editor trying to automatically "fix something".
> 
> 
> Ondrej (온드레이)

Hi Ondrej,

Thanks for this really quick reply.
To answer to your questions: 
- I 'm afraid I have been only able to test it on Debian 10
- I was using ansible 2.10 version but I'm more than happy to test on this other versions.
- Yes sorry it seems that those lines included some blank space that were automatically removed by my editor, should I put them back?

Comment (OndrejHome) [reply README.md]: this change looks OK, resolving

File: README.md (hunk 2 above)
Comment (OndrejHome) [reply README.md]: ok, looks good

File: tasks/cluster_constraint_location.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+---
+- name: Ensure pcs_onstraint_colocation users are present.
```
Comment (OndrejHome) [inline tasks/cluster_constraint_location.yml]: (low) - `pcs_onstraint_colocation users are present.` -> `pcs_constraint_location constraints are present.`

File: tasks/cluster_property.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
+- name: Ensure pcs_property are present.
+  pcs_property:
+    name: "{{ item.name }}"
+    state: "{{ item.state | default(omit) }}"
+    node: "{{ item.node | default(omit) }}"
+    value: "{{ item.value | default(omit) }}"
+    cib_file: "{{ resource.cib_file | default(omit) }}"
```
Comment (OndrejHome) [inline tasks/cluster_property.yml]: (medium) - `resource.cib_file` variable is not defined, most probably this is `item.cib_file`.

File: tasks/cluster_resource.yml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+- name: Ensure pcs_resource are present.
+  pcs_resource:
+    name: "{{ item.name }}"
+    state: "{{ item.state | default(omit) }}"
+    resource_class: "{{ item.resource_class | default(omit) }}"
+    resource_type: "{{ item.resource_type | default(omit) }}"
+    options: "{{ item.options | default(omit) }}"
+    force_resource_update: "{{ item.force_resource_update | default(omit) }}"
+    cib_file: "{{ item.cib_file | default(omit) }}"
+    child_name: "{{ resitemource.child_name | default(omit) }}"
```
Comment (OndrejHome) [inline tasks/cluster_resource.yml]: (medium) - `resitemource.child_name` is not defined, most probably this is `item.child_name`

**Review by OndrejHome — CHANGES_REQUESTED** [review]
Hi @OliPou ,

Sorry for delay here. Testing this after your additional changes there are few issues that needs fixing in code.
As addition I would suggest removing the `cib_file` option which is used when simulating changes in file and not on live cluster. The `cib_file` option here makes a little of sense as main use case is configuring options in the cluster that was setup by role.

If possible please pay more attention to testing the code that you have added to avoid undefined variables. Thank you.

Ondrej (온드레이).

File: tasks/cluster_constraint_location.yml (hunk 3 above)
Comment (OliPou) [reply tasks/cluster_constraint_location.yml]: Sorry fix up on commit :
8cfa853

File: tasks/cluster_property.yml (hunk 4 above)
Comment (OliPou) [reply tasks/cluster_property.yml]: Cib file have been removed on the commit accea57 as you suggest

File: tasks/cluster_resource.yml (hunk 5 above)
Comment (OliPou) [reply tasks/cluster_resource.yml]: Sorry for that must be fixed by the commit 76e6c8f

File: tasks/cluster_constraint_location.yml (hunk 3 above)
Comment (OndrejHome) [reply tasks/cluster_constraint_location.yml]: looks good.

File: tasks/cluster_resource.yml (hunk 5 above)
Comment (OndrejHome) [reply tasks/cluster_resource.yml]: also ok here.

File: tasks/cluster_property.yml (hunk 4 above)
Comment (OndrejHome) [reply tasks/cluster_property.yml]: looks good.

**Comment by OndrejHome** [comment]
Hi @OliPou ,
changes so far looks good. However the commit 22325f5 doesn't look like it should be here at all - it conflicts with master branch where the Debian support was added earlier. To stay within description of this PR I think it will be best to remove/revert it here. It contains some python3 related task that is not present in `master` now.

Apart from that rest looks fairly OK and I will have a look at this once again next week to give it another round of testing.

**Comment by OliPou** [comment]
> Hi @OliPou ,
> changes so far looks good. However the commit [22325f5](https://github.com/OndrejHome/ansible.ha-cluster-pacemaker/commit/22325f5a64eb2643b3069c87e87580d10100d468) doesn't look like it should be here at all - it conflicts with master branch where the Debian support was added earlier. To stay within description of this PR I think it will be best to remove/revert it here. It contains some python3 related task that is not present in `master` now.
> 
> Apart from that rest looks fairly OK and I will have a look at this once again next week to give it another round of testing.

Hi @OndrejHome ,

This commit was add to manage the cluster with python 3, it's just add python3-distutils, which is needed to use your library pcs. Now that python 2 is not longer supported I thought it would have been a good idea to add the python 3 support :) . Tell me if you need me to remove it.

File: tasks/cluster_constraint_colocation.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+---
+- name: Ensure pcs_onstraint_colocation users are present.
```
Comment (tleguern) [inline tasks/cluster_constraint_colocation.yml]:
```suggestion
- name: Ensure pcs_constraint_colocation users are present.
```

File: tasks/cluster_resource_defaults.yml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+---
+- name: Ensure pcs_resource_default are defined.
```
Comment (tleguern) [inline tasks/cluster_resource_defaults.yml]:
```suggestion
- name: Ensure pcs_resource_defaults are defined.
```

File: tasks/cluster_constraint_order.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+---
+- name: Ensure pcs_onstraint_order users are present.
```
Comment (tleguern) [inline tasks/cluster_constraint_order.yml]:
```suggestion
- name: Ensure pcs_constraint_order users are present.
```

**Review by tleguern — COMMENTED** [review]
Hello, I found a few typos while testing the PR.

**Comment by OndrejHome** [comment]
Sorry for delay with further testing here, but due to other things on my side that keeps me busy this would need to wait at least until next week. Thank you for understanding.

Ondrej (온드레이)

**Comment by OndrejHome** [comment]
I have resolved the conflict with main branch where this PR was attempting to add python3 package for debian. Further I have removed the `cib_file` mentioning also from README and unified naming for the tasks that are configuring the various resources added by this PR so it looks more consistent.

Gave it a bit testing and it seems ready for merge. Thank to @tleguern  for review and minor changes and thanks to @OliPou for contribution. In case of future contributions please pay more attention to correctness of documentation (README) as there were significant changes needed to it to make it accurate - I have fixed them.

I will be merging this PR soon.

**Pull request merged**